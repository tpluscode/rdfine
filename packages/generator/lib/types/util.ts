import { GraphPointer } from 'clownface'
import { rdf, rdfs, schema } from '@tpluscode/rdf-ns-builders'
import { TypeMap } from '../types/index.js'

export function isDatatype(term: GraphPointer, typeMappings: TypeMap): boolean {
  if (term.has(rdf.type, rdfs.Datatype).values.length) {
    return true
  }

  const result = term.has(rdf.type, [schema.DataType, rdfs.Datatype]).values.length > 0
  if (result) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isDatatype(superClass, typeMappings)) {
      return true
    }
  }

  return false
}

export function isEnumerationType(term: GraphPointer) {
  if (term.has(rdfs.subClassOf, schema.Enumeration).values.length > 0) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isEnumerationType(superClass)) {
      return true
    }
  }

  return false
}
