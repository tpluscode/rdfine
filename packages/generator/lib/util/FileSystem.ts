import path from 'path'
import { RealFileSystemHost } from '@ts-morph/common'

export default class FileSystem extends RealFileSystemHost {
  private readonly outDir: string

  constructor(outDir: string) {
    super()
    this.outDir = outDir
  }

  getCurrentDirectory(): string {
    return path.resolve(process.cwd(), this.outDir)
  }
}
