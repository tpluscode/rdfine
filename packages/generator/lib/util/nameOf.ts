import { shrink } from '@zazuko/rdf-vocabularies'
import { GraphPointer } from 'clownface'

const rdfineBaseNames = ['equals', 'types', 'pointer', 'unionGraphPointer', '_graphId', '_parent', 'isAnonymous', 'hasType', 'get', 'getArray', 'getBoolean', 'getString', 'getNumber', 'toJSON', '_getObjects', '_create']

export function nameOf(term: GraphPointer, safeIdentifier = false): string {
  const shrunk = shrink(term.value)

  const name = shrunk.split(':')[1]
  if (rdfineBaseNames.includes(name) && safeIdentifier) {
    return `_${name}`
  }
  return name
}
