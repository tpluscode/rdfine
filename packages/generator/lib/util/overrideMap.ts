import { expand } from '@zazuko/rdf-vocabularies'

export function expandMapKeys<T>(overrideMap: Record<string, T>, namespace: string): Record<string, T> {
  return Object.entries(overrideMap)
    .reduce((newMap, [key, override]) => {
      let typeUri: string
      if (key.includes(':')) {
        typeUri = expand(key)
      } else {
        typeUri = `${namespace}${key}`
      }

      return {
        ...newMap,
        [typeUri]: override,
      }
    }, {})
}
