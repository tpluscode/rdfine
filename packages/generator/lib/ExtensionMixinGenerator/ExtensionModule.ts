import { NamespaceDeclarationKind, Project, SourceFile } from 'ts-morph'
import { GraphPointer } from 'clownface'
import { shrink } from '@zazuko/rdf-vocabularies'
import { ExternalResourceType, TypeMetaCollection } from '../types'
import { Context } from '../index'
import { JavascriptProperty } from '../property/JsProperties'
import { PropertyWriter } from '../property/PropertyWriter'
import { MixinModuleBase } from '../MixinGenerator/MixinModuleBase'

export class ExtensionModule extends MixinModuleBase<ExternalResourceType> {
  properties: JavascriptProperty[];
  private extended: { prefix: string; term: string }
  private interfaceName: string

  constructor({ domain, type }: { domain: GraphPointer; type: ExternalResourceType }) {
    super(domain, type)
    this.properties = []
    const [prefix, term] = shrink(this.node.value).split(':')
    this.extended = { prefix, term }
    this.interfaceName = `${this.extended.term}Ex`
  }

  addProperties(...properties: JavascriptProperty[]) {
    for (const property of properties) {
      this.properties.push(property)
    }
  }

  writeModule({ project, context }: { project: Project; types: TypeMetaCollection; context: Omit<Context, 'properties'>; indexModule: SourceFile }): void {
    const module = project.createSourceFile(`extensions/${this.extended.prefix}/${this.extended.term}.ts`, {}, { overwrite: true })

    context.log.debug(`Generating extension mixin for ${this.type.qualifiedName}`)

    const interfaceEx = module.addInterface({
      name: `${this.interfaceName}<D extends RDF.DatasetCore = RDF.DatasetCore>`,
    })

    this.augmentExtendedModule(module)

    const { mixinClass, mixinFunction } = this.createMixinFunction(module, context)

    const propertyWriter = new PropertyWriter({
      interfaceDeclaration: interfaceEx,
      classDeclaration: mixinClass,
      context,
      module: this,
    })
    this.properties.forEach(propertyWriter.addProperty.bind(propertyWriter))

    const nsBuilderTerm = this.extended.term === this.type.localName
      ? `${this.extended.prefix}.${this.extended.term}`
      : `${this.extended.prefix}['${this.extended.term}']`
    module.addStatements([
      `${mixinFunction.getName()}.appliesTo = ${nsBuilderTerm}`,
    ])

    this.addRdfineNamespaceImport(this.type)
    this.addImports(module, context)
  }

  private augmentExtendedModule(module: SourceFile) {
    const augmentedModule = module.addNamespace({
      name: `'${this.type.module}'`,
    }).setDeclarationKind(NamespaceDeclarationKind.Module)

    augmentedModule.addInterface({
      name: `${this.type.localName}<D extends RDF.DatasetCore>`,
      extends: [`${this.interfaceName}<D>`],
    })
  }

  private createMixinFunction(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    const mixinFunction = mixinFile.addFunction({
      name: `${this.type.localName}MixinEx`,
      typeParameters: [{
        name: 'Base',
        constraint: `ExtendingConstructor<${this.type.qualifiedName}, ${this.interfaceName}>`,
      }],
      parameters: [{
        name: 'Resource',
        type: 'Base',
      }],
      returnType: `Constructor<${this.interfaceName} & RdfResourceCore> & Base`,
      isExported: true,
    })

    const mixinClass = mixinFunction.addClass({
      name: 'Impl',
      extends: 'Resource',
      implements: [this.interfaceName],
    })

    mixinClass.addDecorator({
      name: 'namespace',
      arguments: [context.prefix],
    })

    mixinFunction.addStatements('return Impl')

    return { mixinClass, mixinFunction }
  }

  private addImports(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    mixinFile.addImportDeclaration({
      namedImports: ['ExtendingConstructor', 'Constructor', 'namespace', 'property'],
      moduleSpecifier: '@tpluscode/rdfine',
    })
    mixinFile.addImportDeclaration({
      namedImports: ['RdfResourceCore'],
      moduleSpecifier: '@tpluscode/rdfine/RdfResource',
      isTypeOnly: true,
    })

    mixinFile.addImportDeclaration({
      namespaceImport: '$rdf',
      moduleSpecifier: '@rdf-esm/data-model',
    })
    mixinFile.addImportDeclaration({
      namespaceImport: 'RDF',
      moduleSpecifier: 'rdf-js',
      isTypeOnly: true,
    })
    mixinFile.addImportDeclaration({
      namedImports: [this.extended.prefix],
      moduleSpecifier: '@tpluscode/rdf-ns-builders',
    })
    mixinFile.addImportDeclaration({
      namedImports: [context.prefix],
      moduleSpecifier: '../../lib/namespace',
    })
    mixinFile.addImportDeclaration({
      namespaceImport: context.defaultExport,
      moduleSpecifier: '../..',
      isTypeOnly: true,
    })

    Object.entries(this.namespaceImports)
      .forEach(([moduleSpecifier, namespaceImport]) => {
        mixinFile.addImportDeclaration({
          moduleSpecifier,
          namespaceImport,
          isTypeOnly: true,
        })
      })

    this.mixinImports
      .filter(imported => imported.type === 'ExternalResource')
      .forEach(imported => {
        if (imported.type === 'Resource') {
          mixinFile.addImportDeclaration({
            moduleSpecifier: './' + imported.localName,
            namedImports: [imported.mixinName],
          })
        } else {
          const superImport = mixinFile.addImportDeclaration({
            moduleSpecifier: imported.module,
          })

          superImport.addNamedImport({
            name: imported.exportName,
            alias: imported.mixinName,
          })
        }
      })
  }
}
