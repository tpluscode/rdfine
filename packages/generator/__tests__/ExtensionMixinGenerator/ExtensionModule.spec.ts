import { Project, SourceFile } from 'ts-morph'
import cf, { AnyPointer } from 'clownface'
import $rdf from 'rdf-ext'
import { sh } from '@tpluscode/rdf-ns-builders'
import { ExtensionModule } from '../../lib/ExtensionMixinGenerator/ExtensionModule.js'
import { FakeTypeCollection } from '../_helpers/FakeTypeCollection.js'
import { Context } from '../../lib/index.js'
import { fakeLog } from '../_helpers/util.js'
import {expect} from "chai";

describe('ExtensionModule', () => {
  describe('writeModule', () => {
    let project: Project
    let vocabulary: AnyPointer
    let indexModule: SourceFile
    let log: Context['log']

    beforeEach(() => {
      project = new Project({
        useInMemoryFileSystem: true,
      })
      indexModule = project.createSourceFile('index.ts')

      vocabulary = cf({ dataset: $rdf.dataset() })
      log = fakeLog()
    })

    it('creates an extension module',  function () {
      // given
      const module = new ExtensionModule({
        domain: vocabulary.namedNode(sh.PropertyShape),
        type: {
          type: 'ExternalResource',
          localName: 'PropertyShape',
          module: '@rdfine/shacl/lib/PropertyShape',
          mixinName: 'PropertyShapeMixin',
          exportName: '',
          package: '',
          qualifiedMixinName: '',
          qualifiedName: '',
          qualifier: '',
        },
      })

      // when
      module.writeModule({
        project,
        indexModule,
        types: new FakeTypeCollection(),
        context: {
          log,
          defaultExport: 'Dash',
          vocabulary,
          prefix: 'dash',
        },
      })

      // then
      expect(project.getSourceFile('extensions/sh/PropertyShape.ts')).to.matchSnapshot(this)
    })
  })
})
