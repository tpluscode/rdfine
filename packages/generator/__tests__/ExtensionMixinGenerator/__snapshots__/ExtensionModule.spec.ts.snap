// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtensionModule writeModule creates an extension module 1`] = `
import { ExtendingConstructor, Constructor, namespace, property } from "@tpluscode/rdfine";
import type { RdfResourceCore } from "@tpluscode/rdfine/RdfResource";
import * as $rdf from "@rdfjs/data-model";
import type * as RDF from "@rdfjs/types";
import { sh } from "@tpluscode/rdf-ns-builders";
import { dash } from "../../lib/namespace";
import type * as Dash from "../..";
import type * as  from "";

interface PropertyShapeEx<D extends RDF.DatasetCore = RDF.DatasetCore> {
}

declare module '@rdfine/shacl/lib/PropertyShape' {
    interface PropertyShape<D extends RDF.DatasetCore> extends PropertyShapeEx<D> {
    }
}

export function PropertyShapeMixinEx<Base extends ExtendingConstructor<, PropertyShapeEx>>(Resource: Base): Constructor<PropertyShapeEx & RdfResourceCore> & Base {
    @namespace(dash)
    class Impl extends Resource implements PropertyShapeEx {
    }
    return Impl
}
PropertyShapeMixinEx.appliesTo = sh.PropertyShape

`;
